File: backend/onyx/server/manage/prompt_registry.py
from datetime import datetime
from uuid import UUID

from fastapi import APIRouter
from fastapi import Depends
from pydantic import BaseModel
from sqlalchemy.orm import Session

from onyx.auth.users import current_curator_or_admin_user
from onyx.db.engine.sql_engine import get_session
from onyx.db.models import User
from onyx.db.prompt_registry import fetch_prompt_assignments
from onyx.db.prompt_registry import fetch_prompt_template_by_id
from onyx.db.prompt_registry import fetch_prompt_templates
from onyx.db.prompt_registry import fetch_prompt_versions
from onyx.db.prompt_registry import insert_prompt_template
from onyx.db.prompt_registry import insert_prompt_version
from onyx.db.prompt_registry import set_prompt_versions_traffic
from onyx.db.prompt_registry import update_prompt_template
from onyx.db.prompt_registry import upsert_prompt_assignment
from onyx.server.models import StatusResponse

router = APIRouter(prefix="/manage/prompt-registry")


class PromptTemplateCreate(BaseModel):
    name: str
    description: str | None = None


class PromptTemplateUpdate(BaseModel):
    name: str
    description: str | None = None


class PromptTemplateResponse(BaseModel):
    id: int
    name: str
    description: str | None
    owner_user_id: UUID | None
    created_at: datetime
    updated_at: datetime


class PromptVersionCreate(BaseModel):
    content: str


class PromptVersionResponse(BaseModel):
    id: int
    prompt_template_id: int
    version_number: int
    content: str
    created_by_user_id: UUID | None
    created_at: datetime
    is_active: bool
    traffic_percentage: float

class PromptTrafficAllocation(BaseModel):
    version_id: int
    traffic_percentage: float

class PromptTrafficAllocationsRequest(BaseModel):
    allocations: list[PromptTrafficAllocation]


class PromptAssignmentCreate(BaseModel):
    target_type: str
    target_id: str


class PromptAssignmentResponse(BaseModel):
    id: int
    prompt_template_id: int
    target_type: str
    target_id: str
    created_at: datetime


@router.get("/templates")
def get_templates(
    _: User = Depends(current_curator_or_admin_user),
    db_session: Session = Depends(get_session),
) -> list[PromptTemplateResponse]:
    return [
        PromptTemplateResponse(
            id=t.id,
            name=t.name,
            description=t.description,
            owner_user_id=t.owner_user_id,
            created_at=t.created_at,
            updated_at=t.updated_at,
        )
        for t in fetch_prompt_templates(db_session)
    ]


@router.post("/templates")
def create_template(
    template_req: PromptTemplateCreate,
    user: User = Depends(current_curator_or_admin_user),
    db_session: Session = Depends(get_session),
) -> PromptTemplateResponse:
    t = insert_prompt_template(
        name=template_req.name,
        description=template_req.description,
        owner_user_id=user.id,
        db_session=db_session,
    )
    return PromptTemplateResponse(
        id=t.id,
        name=t.name,
        description=t.description,
        owner_user_id=t.owner_user_id,
        created_at=t.created_at,
        updated_at=t.updated_at,
    )


@router.patch("/templates/{template_id}")
def update_template_endpoint(
    template_id: int,
    template_req: PromptTemplateUpdate,
    _: User = Depends(current_curator_or_admin_user),
    db_session: Session = Depends(get_session),
) -> PromptTemplateResponse:
    t = update_prompt_template(
        template_id=template_id,
        name=template_req.name,
        description=template_req.description,
        db_session=db_session,
    )
    return PromptTemplateResponse(
        id=t.id,
        name=t.name,
        description=t.description,
        owner_user_id=t.owner_user_id,
        created_at=t.created_at,
        updated_at=t.updated_at,
    )


@router.delete("/templates/{template_id}")
def delete_template_endpoint(
    template_id: int,
    _: User = Depends(current_curator_or_admin_user),
    db_session: Session = Depends(get_session),
) -> StatusResponse:
    from onyx.db.prompt_registry import delete_prompt_template
    delete_prompt_template(template_id, db_session)
    return StatusResponse(success=True, message="Prompt template deleted.")


@router.get("/templates/{template_id}/versions")
def get_template_versions(
    template_id: int,
    _: User = Depends(current_curator_or_admin_user),
    db_session: Session = Depends(get_session),
) -> list[PromptVersionResponse]:
    return [
        PromptVersionResponse(
            id=v.id,
            prompt_template_id=v.prompt_template_id,
            version_number=v.version_number,
            content=v.content,
            created_by_user_id=v.created_by_user_id,
            created_at=v.created_at,
            is_active=v.is_active,
            traffic_percentage=v.traffic_percentage,
        )
        for v in fetch_prompt_versions(template_id, db_session)
    ]


@router.post("/templates/{template_id}/versions")
def create_template_version(
    template_id: int,
    version_req: PromptVersionCreate,
    user: User = Depends(current_curator_or_admin_user),
    db_session: Session = Depends(get_session),
) -> PromptVersionResponse:
    v = insert_prompt_version(
        template_id=template_id,
        content=version_req.content,
        created_by_user_id=user.id,
        db_session=db_session,
    )
    return PromptVersionResponse(
        id=v.id,
        prompt_template_id=v.prompt_template_id,
        version_number=v.version_number,
        content=v.content,
        created_by_user_id=v.created_by_user_id,
        created_at=v.created_at,
        is_active=v.is_active,
        traffic_percentage=v.traffic_percentage,
    )

@router.post("/templates/{template_id}/traffic")
def update_traffic(
    template_id: int,
    req: PromptTrafficAllocationsRequest,
    _: User = Depends(current_curator_or_admin_user),
    db_session: Session = Depends(get_session),
) -> StatusResponse:
    alloc_dict = {a.version_id: a.traffic_percentage for a in req.allocations}
    set_prompt_versions_traffic(template_id, alloc_dict, db_session)
    return StatusResponse(success=True, message="Traffic allocations updated.")


@router.get("/templates/{template_id}/assignments")
def get_template_assignments(
    template_id: int,
    _: User = Depends(current_curator_or_admin_user),
    db_session: Session = Depends(get_session),
) -> list[PromptAssignmentResponse]:
    return [
        PromptAssignmentResponse(
            id=a.id,
            prompt_template_id=a.prompt_template_id,
            target_type=a.target_type,
            target_id=a.target_id,
            created_at=a.created_at,
        )
        for a in fetch_prompt_assignments(template_id, db_session)
    ]


@router.post("/templates/{template_id}/assignments")
def create_template_assignment(
    template_id: int,
    assignment_req: PromptAssignmentCreate,
    _: User = Depends(current_curator_or_admin_user),
    db_session: Session = Depends(get_session),
) -> PromptAssignmentResponse:
    a = upsert_prompt_assignment(
        template_id=template_id,
        target_type=assignment_req.target_type,
        target_id=assignment_req.target_id,
        db_session=db_session,
    )
    return PromptAssignmentResponse(
        id=a.id,
        prompt_template_id=a.prompt_template_id,
        target_type=a.target_type,
        target_id=a.target_id,
        created_at=a.created_at,
    )

File: web/src/lib/prompt-registry.ts
import useSWR from "swr";
import { errorHandlingFetcher } from "@/lib/fetcher";

export interface PromptTemplate {
  id: number;
  name: string;
  description: string | null;
  owner_user_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface PromptVersion {
  id: number;
  prompt_template_id: number;
  version_number: number;
  content: string;
  created_by_user_id: string | null;
  created_at: string;
  is_active: boolean;
  traffic_percentage: number;
}

export interface PromptAssignment {
  id: number;
  prompt_template_id: number;
  target_type: string;
  target_id: string;
  created_at: string;
}

export function usePromptTemplates() {
  const { data, error, isLoading, mutate } = useSWR<PromptTemplate[]>(
    "/api/manage/prompt-registry/templates",
    errorHandlingFetcher
  );
  return {
    templates: data || [],
    isLoading,
    error,
    refresh: mutate,
  };
}

export function usePromptVersions(templateId: number) {
  const { data, error, isLoading, mutate } = useSWR<PromptVersion[]>(
    `/api/manage/prompt-registry/templates/${templateId}/versions`,
    errorHandlingFetcher
  );
  return {
    versions: data || [],
    isLoading,
    error,
    refresh: mutate,
  };
}

export function usePromptAssignments(templateId: number) {
  const { data, error, isLoading, mutate } = useSWR<PromptAssignment[]>(
    `/api/manage/prompt-registry/templates/${templateId}/assignments`,
    errorHandlingFetcher
  );
  return {
    assignments: data || [],
    isLoading,
    error,
    refresh: mutate,
  };
}

export async function createPromptTemplate(name: string, description?: string) {
  const res = await fetch("/api/manage/prompt-registry/templates", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name, description }),
  });
  if (!res.ok) throw new Error("Failed to create prompt template");
  return res.json();
}

export async function createPromptVersion(templateId: number, content: string) {
  const res = await fetch(`/api/manage/prompt-registry/templates/${templateId}/versions`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ content }),
  });
  if (!res.ok) throw new Error("Failed to create prompt version");
  return res.json();
}

export async function setPromptTrafficAllocations(templateId: number, allocations: { version_id: number; traffic_percentage: number }[]) {
  const res = await fetch(`/api/manage/prompt-registry/templates/${templateId}/traffic`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ allocations }),
  });
  if (!res.ok) throw new Error("Failed to set traffic allocations");
  return res.json();
}

export async function createPromptAssignment(templateId: number, targetType: string, targetId: string) {
  const res = await fetch(`/api/manage/prompt-registry/templates/${templateId}/assignments`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ target_type: targetType, target_id: targetId }),
  });
  if (!res.ok) throw new Error("Failed to create prompt assignment");
  return res.json();
}

export async function deletePromptTemplate(templateId: number) {
  const res = await fetch(`/api/manage/prompt-registry/templates/${templateId}`, {
    method: "DELETE",
  });
  if (!res.ok) throw new Error("Failed to delete prompt template");
  return res.json();
}
